src/NodaTime/YearMonth.cs

```diff
@@ -0,0 +1,272 @@
+    /// almost always indicates a bug in the calling code.
+    /// </para>
+    /// </remarks>
+    /// <threadsafety>This type is an immutable value type. See the thread safety section of the user guide for more information.</threadsafety>
+    public struct YearMonth : IEquatable<YearMonth>, IComparable<YearMonth>, IComparable, IFormattable, IXmlSerializable
+    {
+        /// <summary>
+        /// The start of month. This is used as our base representation as we already have
+        /// plenty of other code that integrates it, and it handles the compact representation.
+        /// </summary>
+        private readonly YearMonthDayCalendar startOfMonth;
+
+        /// <summary>Gets the calendar system associated with this year/month.</summary>
+        /// <value>The calendar system associated with this year/month.</value>
+        public CalendarSystem Calendar => CalendarSystem.ForOrdinal(startOfMonth.CalendarOrdinal);
+
+        /// <summary>Gets the year of this year/month.</summary>
+        /// <remarks>This returns the "absolute year", so, for the ISO calendar,
+        /// a value of 0 means 1 BC, for example.</remarks>
+        /// <value>The year of this year/month.</value>
+        public int Year => startOfMonth.Year;
+
+        /// <summary>Gets the month of this year/month within the year.</summary>
+        /// <value>The month of this year/month within the year.</value>
+        public int Month => startOfMonth.Month;
+
+        /// <summary>Gets the year of this value within the era.</summary>
+        /// <value>The year of this value within the era.</value>
+        public int YearOfEra => Calendar.GetYearOfEra(startOfMonth.Year);
+
+        /// <summary>Gets the era of this year/month.</summary>
+        /// <value>The era of this year/month.</value>
+        public Era Era => Calendar.GetEra(startOfMonth.Year);
+
+        /// <summary>
+        /// Returns the date of the start of this year/month.
+        /// </summary>
+        public LocalDate StartDate => new LocalDate(startOfMonth);
+
+        /// <summary>
+        /// Returns the date of the end of this year/month.
+        /// </summary>
+        public LocalDate EndDate => new LocalDate(Year, Month, Calendar.GetDaysInMonth(Year, Month), Calendar);
+
+        private YearMonthDay YearMonthDay => startOfMonth.ToYearMonthDay();
+
+        /// <summary>
+        /// Constructs an instance for the given year and month in the ISO calendar.
+        /// </summary>
+        /// <param name="year">The year. This is the "absolute year", so a value of 0 means 1 BC, for example.</param>
+        /// <param name="month">The month of year.</param>
+        /// <returns>The resulting year/month.</returns>
+        /// <exception cref="ArgumentOutOfRangeException">The parameters do not form a valid year/month.</exception>
+        public YearMonth(int year, int month)
+        {
+            if (year < MinGregorianYear || year > MaxGregorianYear || month < 1 || month > 12)
```

Comment:
I'm assuming you're not delegating to the `(int year, int month, CalendarSystem)` constructor for performance reasons? If so, add a comment?

Reply:
Done.